src/main/java/net/minecraftforge/common/data/JsonCodecProvider.java

```diff
@@ -0,0 +1,103 @@
+{
+    private static final Logger LOGGER = LogManager.getLogger();
+    
+    public JsonCodecProvider
+    {
+        // Track generated data so other dataproviders can validate if needed.
+        final ResourceType resourceType = new ResourceType(packType, ".json", directory);
+        for (ResourceLocation id : entries.keySet())
+        {
+            existingFileHelper.trackGenerated(id, resourceType);
+        }
+    }
+    
+    /**
+     * {@return DatapackRegistryProvider that encodes using the registered loading codec for the provided registry key}
+     * Ensures the correct directory and codec are used.
+     * Only vanilla datapack registries enumerated in {@link RegistryAccess} and custom forge datapack registries can
+     * be generated this way.
+     * 
+     * @param <T> Registry element type, e.g. Biome
+     * @param dataGenerator DataGenerator provided by {@link GatherDataEvent}.
+     * @param modid namespace of the mod adding this DataProvider, for logging purposes.
+     * @param registryOps RegistryOps to encode values to json with.
+     * @param registryKey ResourceKey identifying the registry and its directory.
+     * @param entries Map of entries to encode and their ResourceLocations. Paths for values are derived from the ResourceLocation's entryid:entrypath.
+     */
+    public static <T> JsonCodecProvider<T> forDatapackRegistry(DataGenerator dataGenerator, ExistingFileHelper existingFileHelper, String modid, RegistryOps<JsonElement> registryOps, ResourceKey<Registry<T>> registryKey, Map<ResourceLocation, T> entries)
+    {
+        final ResourceLocation registryId = registryKey.location();
+        // Minecraft datapack registry folders are in data/json-namespace/registry-name/
+        // Non-vanilla registry folders are data/json-namespace/registry-namespace/registry-name/
+        final String registryFolder = registryId.getNamespace().equals("minecraft")
+            ? registryId.getPath()
+            : registryId.getNamespace() + "/" + registryId.getPath();
+        final Codec<T> codec = (Codec<T>)RegistryAccess.REGISTRIES.get(registryKey).codec();
+        return new JsonCodecProvider<>(dataGenerator, existingFileHelper, modid, registryOps, PackType.SERVER_DATA, registryFolder, codec, entries);
+    }
+
+    @Override
+    public void run(final CachedOutput cache) throws IOException
+    {
+        final Path outputFolder = this.dataGenerator().getOutputFolder();
+        final String dataFolder = this.packType().getDirectory();
+        for (final var entry : this.entries().entrySet())
+        {
+            final ResourceLocation id = entry.getKey();
+            final T value = entry.getValue();
+            final Path path = outputFolder.resolve(String.join("/", dataFolder, id.getNamespace(), this.directory(), id.getPath()+".json"));
```

Comment:
Formatting: spaces around `+` operator.
```suggestion
            final Path path = outputFolder.resolve(String.join("/", dataFolder, id.getNamespace(), this.directory(), id.getPath() + ".json"));
```